crates/counter-agg/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+use time_weighted_average::tspoint::TSPoint;
+use serde::{Deserialize, Serialize};
+use regression::{XYPair, RegressionSummary};
+
+pub mod regression;
+pub mod range;
+mod tests;
+
+#[derive(Debug, PartialEq)]
+pub enum CounterError{
+    OrderError,
+    BoundsInvalid,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct CounterSummary {
+    pub first: TSPoint,
+    pub second: TSPoint,
+    pub penultimate: TSPoint,
+    pub last: TSPoint,
+    pub reset_sum: f64,
+    pub num_resets: u64,
+    pub num_changes: u64,
+    pub regress: RegressionSummary,
+    pub bounds: Option<range::I64Range>,
+}
+
+// Note that this can lose fidelity with the timestamp, but it would only lose it in the microseconds, 
+// this is likely okay in most applications. However, if you need better regression analysis at the subsecond level, 
+// you can always subtract a common near value from all your times, then add it back in, the regression analysis will be unchanged.
+// TODO Should this be a method?
+fn ts_to_xy(pt: TSPoint) -> XYPair{
+    XYPair{
+        x: pt.ts as f64,
+        y: pt.val,
+    }
+}
+
+pub fn to_seconds(t: f64)-> f64{
+    t / 1_000_000 as f64// by default postgres timestamps have microsecond precision
+}
+
+impl CounterSummary {
+    pub fn new(pt: &TSPoint, bounds:Option<range::I64Range>) -> CounterSummary {
+        let mut n = CounterSummary{
+            first: *pt,
+            second: *pt,
+            penultimate: *pt,
+            last: *pt,
+            reset_sum: 0.0,
+            num_resets: 0,
+            num_changes: 0,
+            regress: RegressionSummary::new(),
+            bounds,
+        };
+        n.regress.accum(ts_to_xy(*pt)).unwrap();
+        n
+    }
+
+    // expects time-ordered input 
+    pub fn add_point(&mut self, incoming: &TSPoint) -> Result<(), CounterError>{
+
+        if incoming.ts < self.last.ts {
+            return Err(CounterError::OrderError);
+        }
+        if incoming.val < self.last.val {
+            self.reset_sum += self.last.val;
+            self.num_resets+= 1;
+        }
+        // right now we treat a counter reset that goes to exactly zero as a change (not sure that's correct, but it seems defensible)
+        if incoming.val != self.last.val{
+            self.num_changes += 1;
+        }
+        if self.first == self.second {
+            self.second = *incoming;
+        }
+        self.penultimate = self.last;
+        self.last = *incoming;
+        let mut incoming_xy = ts_to_xy(*incoming);
+        incoming_xy.y += self.reset_sum;
+        self.regress.accum(incoming_xy).unwrap();
+        Ok(())
+    }
+
+    fn single_value(&self) -> bool {
+        self.last == self.first
+    }
+
+    // combining can only happen for disjoint time ranges 
+    pub fn combine(&mut self, incoming: &CounterSummary) -> Result<(), CounterError> {
```

Comment:
I think `append` better captures the behavior and requirements of this function.

Reply:
Hmmm...I feel like append seems like you're just pushing another into a list of them or something, combine seems like it works better for me, I suppose maybe `ordered_combine` or something. I do see where you're coming from though, I just think append is worse, but I think there's probably a better name than either of them out there...I just can't think of it :)